lessons/4.mdx

```diff
@@ -0,0 +1,818 @@
+      )
+    })
+
+    it('should return proper finalJSON for Tier 1', async function () {
+      await expect(
+        await contract.finalJSON('TierNFT', 222, 'IMAGE_SVG_BASE64_HERE', 1),
+      ).to.equal(
+        '{"name": "TierNFT #222", "description": "TierNFTs collection", "image": "data:image/svg+xml;base64,IMAGE_SVG_BASE64_HERE","attributes":[{"trait_type": "Tier", "value": "Medium" }]}',
+      )
+    })
+
+    it('should return proper finalJSON for Tier 2', async function () {
+      await expect(
+        await contract.finalJSON('TierNFT', 333, 'IMAGE_SVG_BASE64_HERE', 2),
+      ).to.equal(
+        '{"name": "TierNFT #333", "description": "TierNFTs collection", "image": "data:image/svg+xml;base64,IMAGE_SVG_BASE64_HERE","attributes":[{"trait_type": "Tier", "value": "Premium" }]}',
+      )
+    })
+  })
+
+  it('should return base64-encoded data for tokenURI for Tier 0', async function () {
+    await contract.mint({
+      value: hre.ethers.utils.parseEther('0.01'),
+    })
+    await expect(await contract.tokenURI(1)).to.equal(
+      'data:application/json;base64, <full string removed to fit in lesson>',
+    )
+  })
+})
+```
+
+The final test is an example mentioned earlier about adding a test before
+refactoring with helper methods. It checks that the initial `tokenURI()` method
+(which returns an encoded Base64 string) returns the right result. It's helpful
+while refactoring to make sure we didn't break things along the way, and it's
+helpful at the end to ensure the `tokenURI()` method continues to work without
+fail when code changes.
+
+## Food for Thought
+
+With all the helpfulness of tests, why didn't we start with tests before we
+wrote any of the contract implementation?
+
+The main reason was trying to minimize all the things you needed to learn to get
+started, but the methodology when writing tests first is called "Test Driven
+Development" (TDD).
+
+TDD is a very powerful methodology which goes like this:
+
+- before you write implementation code, you write one or more tests and run
+  them.
+- of course those tests fail initially since there are methods that don't have
+  any implementaiton yet. As mentioned earlier, it's always important to prove
+  your test fails initially.
+- then add your implementation until the tests pass.
+
+Voila!
+
+With TDD:
+
+- you would have written tests as you go, which are much more fun to write than
+  all at once at the end
+- you would have ended up writing better code because you discover quickly when
+  code is impossible to test. You'll end up with added benefit of breaking a big
+  method into multiple smaller methods that are easily testable. So in the
+  refactor example above for `tokenURI()` you would never have written the
+  version that outputs a long base64 string because it's not testable.
+
+  Bolting on tests after the fact can be painful, boring and you lose the
```

Comment:
Bolting on tests after the fact can be painful and boring, and you lose the potential code improvements you can make when you test as you go.
*after the fact* ..... never heard it before, but I like it!!

Brilliant lesson @briangershon

Reply:
Thanks @elPiablo I really appreciate your perspective and all the feedback on this lesson!